Integrate PeerDAS cryptography library

kevaundray opened the pull request:
<!-- Thanks for sending a pull request! Please check out our contribution guidelines: -->
<!-- https://github.com/ConsenSys/teku/blob/master/CONTRIBUTING.md -->

## PR Description

On top of c-kzg, we plan to have a Rust library that compiles to Java and all of the other targets. A preliminary release has been made, this PR is being made to show the end to end integration of the library and possibly surface any issues.

## Fixed Issue(s)
<!-- Please link to fixed issue(s) here using format: fixes #<issue number> -->
<!-- Example: "fixes #2" -->

## Documentation

- [ ] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [ ] I thought about adding a changelog entry, and added one if I deemed necessary.

kevaundray committed:
chore: add initial library for integration test

kevaundray commented:
Note: Things like namespaces will change in the future

kevaundray committed:
chore: add import

kevaundray commented:
One thing to note:

- This repo currently uses an old API for c-kzg where you are able to recover the Blob. The new consensus-specs does not allow this, it has become encapsulated in a method called recoverCellsAndproofs which returns all of the cells and the proofs for them.

kevaundray commented:
Noting the use of BeforeEach for the setup method in tests. This will implicitly test that loading the same trusted setup does not throw or break. This is in addition to the explicit `testKzgLoadSameTrustedSetupTwice_shouldNotThrowException` function.

See here: https://github.com/Nashatyrev/teku/blob/3768eca8cf0b8bc5989091a9c8626ddc4a481c24/infrastructure/kzg/src/test/java/tech/pegasys/teku/kzg/CKZG4844Test.java#L53

*BeforeEach as opposed to BeforeAll

kevaundray committed:
update lib to version 0.0.4:

- change namespace and ensres that the library is loaded once

kevaundray committed:
Add a conditional compile on PeerDASKZG and integrate it into computeCells

kevaundray commented:
Currently to run tests with the rust lib, you can do `./gradlew :infrastructure:kzg:testWithPeerDas` , the end solution may look something like ./gradlew test -D use.peer.das not really sure what the teku convention is here, so open to whatever is best

kevaundray committed:
chore: modify computeCellsAndProofs

kevaundray committed:
chore: pull out duplicated code in `computeCellsAndProofs`

kevaundray committed:
chore: modify verifyCellProof

kevaundray committed:
chore: modify verifyCellKZGProofBatch

kevaundray committed:
chore: modify recoverCells

kevaundray commented:
Something I noticed, some methods have try-catch logic while others (mainly the new peerdas methods) do not -- Not blocking for this PR, though it looks undesirable since you are not wrapping the potentially arbitrary exceptions from the underlying crypto library into a KZGException

kevaundray committed:
chore: update to 0.0.6 (synchronized native lib loading)

kevaundray commented:
We could also modify BlobToKZGCommitment to:

```java
@Override
  public KZGCommitment blobToKzgCommitment(final Bytes blob) throws KZGException {

    try {
      byte[] commitmentBytes;

      if (USE_PEER_DAS) {
        commitmentBytes = peerDASinstance.blobToKZGCommitment(blob.toArrayUnsafe());
      } else {
        commitmentBytes = CKZG4844JNI.blobToKzgCommitment(blob.toArrayUnsafe());
      }

      return KZGCommitment.fromArray(commitmentBytes);
    } catch (final Exception ex) {
      throw new KZGException("Failed to produce KZG commitment from blob", ex);
    }
  }
  ```
  
  But there are CKZG specific errors in the tests so that would require more changes than I am hoping for

kevaundray committed:
chore: update to 0.0.8

jtraglia commented on infrastructure/kzg/src/main/java/tech/pegasys/teku/kzg/CKZG4844.java:
```diff
@@ -34,6 +38,8 @@ final class CKZG4844 implements KZG {
 
   private static CKZG4844 instance;
+  private static LibPeerDASKZG peerDASinstance;
+  private static final boolean USE_PEER_DAS = Boolean.getBoolean("use.peer.das");
```
Personally, I think `USE_PEER_DAS` is a bit too ambiguous.

jtraglia commented on infrastructure/kzg/src/main/java/tech/pegasys/teku/kzg/CKZG4844.java:
```diff
@@ -62,6 +68,7 @@ public synchronized void loadTrustedSetup(final String trustedSetupFile) throws
       return;
     }
     try {
+      peerDASinstance = new LibPeerDASKZG();
```
Is it possible to only initialize this if `USE_PEER_DAS` is true?

zilm13 committed:
Merge branch 'das' into fork/kevaundray/kw/peerdas-kzg

zilm13 committed:
Fix recoverCellsAndProofs

kevaundray committed:
update lib to 0.0.9

kevaundray committed:
chore(multi):

- update FLAG name from USE_PEER_DAS to `USE_RUST_KZG_LIB`
- We now initialize and free the rust kzg lib when the flag is enabled

kevaundray commented:
Changed the flag name, tests can now be ran with `./gradlew :infrastructure:kzg:testWithRustKZG`

zilm13 committed:
Refactor RustKZG to use separate class with delegate

zilm13 committed:
Init RustKZG only on loadTrustedSetup

zilm13 committed:
Add --Xrust-kzg-enabled param to enable rust kzg for PeerDAS

zilm13 commented:
I've finished with all changes I wanted, happy to merge.
So at the end we have `--Xrust-kzg-enabled` command line option to enable library and separate tests for both implementation and rust with fallback variant, which will be used with flag.

zilm13 reviewed (approved):
LGTM

kevaundray commented:
> I've finished with all changes I wanted, happy to merge. So at the end we have `--Xrust-kzg-enabled` command line option to enable library and separate tests for both implementation and rust with fallback variant, which will be used with flag. I've tested with flag enabled in Kurtosis, no issues detected.

Thanks for doing the heavy lifting to separate the two libraries into two different files and adding a flag to the command line!
